File: csharp/1089.duplicate-zeros.cs
/*
 * @lc app=leetcode id=1089 lang=csharp
 *
 * [1089] Duplicate Zeros
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {
    public void DuplicateZeros(int[] arr) {
        bool flag = false;
        int n = arr.Length;
        var queue = new Queue<int>();
        for (int i = 0; i < n; ++i) {
            queue.Enqueue(arr[i]);
            if (flag) {
                arr[i] = 0;
                flag = false;
                continue;
            }
            int c = queue.Dequeue();
            if (c == 0) {
                flag = true;
            }
            arr[i] = c;
        }
    }
}
// @lc code=end

File: java/1089.duplicate-zeros.java
/*
 * @lc app=leetcode id=1089 lang=java
 *
 * [1089] Duplicate Zeros
 */

import java.util.ArrayDeque;
import java.util.Queue;

// @lc code=start

class Solution {
    public void duplicateZeros(int[] arr) {
        int n = arr.length;
        Queue<Integer> queue = new ArrayDeque<>(n);
        boolean dup = false;
        for (int i = 0; i < n; ++i) {
            queue.offer(arr[i]);
            if (dup) {
                dup = false;
                arr[i] = 0;
                continue;
            }
            int x = queue.poll();
            dup = x == 0;
            arr[i] = x;
        }
    }
}
// @lc code=end

